### PR TITLE
Update the iris sklearn/xgboost sample

### PR DESCRIPTION
**Pull request by 0olwzo0**
1. Fix the bugs in the code. e.g. `bucket` is not defined.
2. Introduce `--job-dir` as output model dir, which is the common practice for CMLE training.
3. Split the input data to 'train' and 'test' sets.  And add model evaluation after train.
3. Add hptuning support using `cloudml-hypertune`.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/cloudml-samples/229)

### CODE REVIEW

File: sklearn/README.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  --module-name $MAIN_TRAINER_MODULE \
+  --region us-central1 \
+  --runtime-version $RUNTIME_VERSION \
+  --python-version $PYTHON_VERSION \
+  --scale-tier BASIC \
+  --config hyperparam.yaml
+```
+
+### Online Predcition
+If you have done with tuning a model and hope to serve the model for online prediction, please follow instruciton in the [Online Prediction with scikit-learn on Google Cloud Machine Learning Engine](https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/sklearn/notebooks/Online%20Prediction%20with%20scikit-learn.ipynb).
```
Comment (nnegrey) [inline sklearn/README.md]: Once you are done training / tuning your model. Check out our guides on how to serve the model for online prediction.
* [Online Prediction with scikit-learn on Google Cloud Machine Learning Engine](https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/sklearn/notebooks/Online%20Prediction%20with%20scikit-learn.ipynb).
* [Getting Started with scikit-learn and XGBoost online predictions](https://cloud.google.com/ml-engine/docs/scikit/quickstart)
* [Online Predictions with scikit-learn Pipelines](https://cloud.google.com/ml-engine/docs/scikit/using-pipelines)

File: sklearn/iris_sklearn_trainer/iris.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # This file is for training on Cloud ML Engine with scikit-learn.
 
-
 # [START setup]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [START mle_iris_training_setup]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 3]
```diff
@@ -10,11 +10,29 @@
 
 from sklearn import svm
 from sklearn.externals import joblib
+from sklearn.model_selection import train_test_split
 
-# Fill in your Cloud Storage bucket name
-BUCKET_ID = <YOUR_BUCKET_NAME>
-# [END setup]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [END mle_iris_training_setup]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 4]
```diff
@@ -10,11 +10,29 @@
...
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    '--job-dir',
+    help='GCS location to write checkpoints and export models',
+    required=True
+)
+parser.add_argument(
+    '--kernel',
+    help='Specifies the kernel type to be used in the '
+         'algorithm.It must be one of `linear`, `poly`, '
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: add space between "algorithm.It"

File: sklearn/iris_sklearn_trainer/iris.py [hunk 5]
```diff
@@ -10,11 +10,29 @@
...
+         '`rbf`, `sigmoid` or a callable',
+    default='rbf'
+)
+parser.add_argument(
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: Space between

File: sklearn/iris_sklearn_trainer/iris.py [hunk 6]
```diff
@@ -10,11 +10,29 @@
...
+    '--c',
+    help='Penalty parameter C of the error term',
+    type=float,
+    default='1.0'
+)
 
+args = parser.parse_args()
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: parsing should be inside a `if __name__ == '__main__':` statement at the bottom of the python file.

File: sklearn/iris_sklearn_trainer/iris.py [hunk 7]
```diff
@@ -10,11 +10,29 @@
...
 
 # [START download-data]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [START mle_iris_training_download-data]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 8]
```diff
@@ -30,29 +48,41 @@
                        iris_target_filename], stderr=sys.stdout)
 # [END download-data]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [END mle_iris_training_download-data]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 9]
```diff
@@ -30,29 +48,41 @@
...
 
-
 # [START load-into-pandas]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [START mle_iris_training_load-data]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 10]
```diff
@@ -30,29 +48,41 @@
...
 # Load data into pandas
 iris_data = pd.read_csv(iris_data_filename).values
 iris_target = pd.read_csv(iris_target_filename).values
 iris_target = iris_target.reshape((iris_target.size,))
+# Split the input data into training and test sets.
+# A fixed random_state will always return the same split result.
+train_data, test_data, train_target, test_target = train_test_split(
+    iris_data, iris_target, test_size=0.4, random_state=0)
 # [END load-into-pandas]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [END mle_iris_training_load-data]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 11]
```diff
@@ -30,29 +48,41 @@
...
 
-
 # [START train-and-save-model]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [START mle_iris_training_train_save_model]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 12]
```diff
@@ -30,29 +48,41 @@
...
 # Train the model
-classifier = svm.SVC(verbose=True)
-classifier.fit(iris_data, iris_target)
-
+classifier = svm.SVC(verbose=True, kernel=args.kernel, C=args.c)
+classifier.fit(train_data, train_target)
 # Export the classifier to a file
 model = 'model.joblib'
 joblib.dump(classifier, model)
 # [END train-and-save-model]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [END mle_iris_training_train_save_model]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 13]
```diff
@@ -30,29 +48,41 @@
...
 
+# [START report-hyperparameter-tuning-metrics]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [START mle_iris_training_hp_tuning_metrics]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 14]
```diff
@@ -30,29 +48,41 @@
...
+# Calculate the mean accuracy on the given test data and labels.
+score = classifier.score(test_data, test_target)
+# Report the mean accuracy as hyperparameter tuning objective metric.
+import hypertune
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: Import at top of file

File: sklearn/iris_sklearn_trainer/iris.py [hunk 15]
```diff
@@ -30,29 +48,41 @@
...
+hpt = hypertune.HyperTune()
+hpt.report_hyperparameter_tuning_metric(
+    hyperparameter_metric_tag='my_metric_tag',
+    metric_value=score,
+    global_step=1)
+# [END report-hyperparameter-tuning-metrics]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [END mle_iris_training_hp_tuning_metrics]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 16]
```diff
@@ -30,29 +48,41 @@
...
 
 # [START upload-model]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [START mle_iris_training_upload_model]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 17]
```diff
@@ -30,29 +48,41 @@
...
 # Upload the saved model file to Cloud Storage
-model_path = os.path.join('gs://', bucket, datetime.datetime.now().strftime(
-    'iris_%Y%m%d_%H%M%S'), model)
+model_path = os.path.join(args.job_dir, model)
 subprocess.check_call(['gsutil', 'cp', model, model_path], stderr=sys.stdout)
 # [END upload-model]
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: `# [END mle_iris_training_upload_model]`

File: sklearn/iris_sklearn_trainer/iris.py [hunk 18]
```diff
@@ -30,29 +48,41 @@
...
+    metric_value=score,
+    global_step=1)
+# [END report-hyperparameter-tuning-metrics]
 
 # [START upload-model]
 # Upload the saved model file to Cloud Storage
-model_path = os.path.join('gs://', bucket, datetime.datetime.now().strftime(
-    'iris_%Y%m%d_%H%M%S'), model)
+model_path = os.path.join(args.job_dir, model)
 subprocess.check_call(['gsutil', 'cp', model, model_path], stderr=sys.stdout)
```
Comment (nnegrey) [inline sklearn/iris_sklearn_trainer/iris.py]: Please update to use the python google cloud storage package: `from google.cloud import storage`

Similar to this flow from: https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/sklearn/notebooks/ml_engine_training.ipynb
```
# Upload the model to GCS
bucket = storage.Client().bucket(BUCKET_ID)
blob = bucket.blob('{}/{}'.format(
    datetime.datetime.now().strftime('census_%Y%m%d_%H%M%S'),
    model))
blob.upload_from_filename(model)
```

File: sklearn/setup.py [hunk 19]
```diff
@@ -0,0 +1,14 @@
+from setuptools import find_packages
```
Comment (nnegrey) [inline sklearn/setup.py]: Is this file being used?
I think we're also trying to move away from this format of a setup.py file, but let me double check.

**Comment by 0olwzo0** [comment]
@dizcology , @nnegrey, @thedriftofwords  We have customers asked about hptuning support of sklearn/xgboost on stackoverflow(https://stackoverflow.com/questions/52662854/hyperparameters-tuning-with-google-cloud-ml-engine-and-xgboost). Could we publish this sample as a reference before we have it well documented in CMLE docs?

File: sklearn/iris/README.md [hunk 20]
```diff
@@ -0,0 +1,105 @@
+# Training a scikit-learn model on Cloud ML Engine using the Iris dataset
+
+- - -
+
+### Setup GCP environment
+Please fllow this [instruction](https://cloud.google.com/ml-engine/docs/scikit/getting-started-training#before_you_begin) to set up your environment with scikit-learn and your Google Cloud Project to train your model on Google Cloud ML Engine.
```
Comment (thebinaryone1) [inline sklearn/iris/README.md]: follow

Comment (thebinaryone1) [inline sklearn/iris/README.md]: these

Comment (thebinaryone1) [inline sklearn/iris/README.md]: follow these instructions

File: sklearn/iris/README.md [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+
+### Install required packages for training in local environment.
+In order to run the sample in your local enviroment, you need to
+run the following commands to install required python packages.
```
Comment (thebinaryone1) [inline sklearn/iris/README.md]: command

File: sklearn/iris/README.md [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+  --job-dir $JOB_DIR \
+  --package-path $TRAINING_PACKAGE_PATH \
+  --module-name $MAIN_TRAINER_MODULE \
+  --region $REGION \
+  --runtime-version=$RUNTIME_VERSION \
+  --python-version=$PYTHON_VERSION \
+  --scale-tier $SCALE_TIER
+```
+### Tuning Hyperparameters on Cloud ML Engine
+You're also able to do hyperparameter tuning with scikit-learn on Cloud ML Engine. The python package [cloudml-hypertune](https://pypi.org/project/cloudml-hypertune/) is used to help you report hyperparameter tuning objective metrics when running on Cloud ML Engine. cloudml-hypertune can work with any ML framework.
```
Comment (thebinaryone1) [inline sklearn/iris/README.md]: Aren't these the same?

File: xgboost/iris/README.md [hunk 23]
```diff
@@ -0,0 +1,96 @@
+# Training an XGBoost model on Cloud ML Engine using the Iris dataset
+
+- - -
+
+### Setup GCP environment
+Please fllow this [instruction](https://cloud.google.com/ml-engine/docs/scikit/getting-started-training#before_you_begin) to set up your environment with xgboost and your Google Cloud Project to train your model on Google Cloud ML Engine.
```
Comment (thebinaryone1) [inline xgboost/iris/README.md]: ditto.

File: xgboost/iris/README.md [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+
+
+### Install required packages for training in local environment.
+In order to run the sample in your local enviroment, you need to
+run the following commands to install required python packages.
```
Comment (thebinaryone1) [inline xgboost/iris/README.md]: command

File: xgboost/iris/README.md [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+  --module-name $MAIN_TRAINER_MODULE \
+  --region $REGION \
+  --runtime-version=$RUNTIME_VERSION \
+  --python-version=$PYTHON_VERSION \
+  --scale-tier $SCALE_TIER
+```
+### Tuning Hyperparameters on Cloud ML Engine
+You're also able to do hyperparameter tuning with xgboost on Cloud ML Engine. The python package [cloudml-hypertune](https://pypi.org/project/cloudml-hypertune/) is used to help you report hyperparameter tuning objective metrics when running on Cloud ML Engine. cloudml-hypertune can work with any ML framework.
+
+You're also able to do hyperparameter tuning with Scikit-Learn on Cloud ML Engine. We introduce a new python package [cloudml-hypertune](https://pypi.org/project/cloudml-hypertune/) to help you report hyperparameter tuning objective metrics when running on Cloud ML Engine. It can work with any ML framework.
```
Comment (thebinaryone1) [inline xgboost/iris/README.md]: Don't need this, right? Since we have a separate README for sklearn.

File: sklearn/iris/README.md [hunk 26]
```diff
@@ -0,0 +1,103 @@
+# Training a scikit-learn model on Cloud ML Engine using the Iris dataset
+
+- - -
```
Comment (dizcology) [inline sklearn/iris/README.md]: maybe say a few words about what this sample is about here.

File: sklearn/iris_training.py [hunk 27]
```diff
@@ -1,5 +1,6 @@
 # This file is for training on Cloud ML Engine with scikit-learn.
-
+# TODO(0olwzo0): Remove this file after stopping referring this file from CMLE
```
Comment (dizcology) [inline sklearn/iris_training.py]: is this file still used?

**Comment by gogasca** [comment]
Thanks for the contribution, any update on the changes?

**Comment by andrewferlitsch** [comment]
@nnegrey This seems kind of old. Can the last change be made and approved/merged? Or is the sample still needed?

**Comment by nnegrey** [comment]
@thedriftofwords is this used in docs?

**Comment by andrewferlitsch** [comment]
@nnegrey I concur this is too old. Let's close this w/o merging
